core/src/main/java/hudson/util/TextFile.java

```diff
@@ -113,7 +112,7 @@ public void write(String text) throws IOException {
     public @NonNull String head(int numChars) throws IOException {
         char[] buf = new char[numChars];
         int read = 0;
-        try (Reader r = new FileReader(file)) {
+        try (Reader r = Files.newBufferedReader(Util.fileToPath(file), Charset.defaultCharset())) {
```

Comment:
`TextFileTest` uses `Charset.defaultCharset` when writing the test string, so it seems only reasonable that we would also use it when reading the same string.

Reply:
But it does match https://github.com/jenkinsci/jenkins/blob/75ce81d12ca29eb4a3bc7bd3a4018d65452e90d6/core/src/main/java/hudson/util/TextFile.java#L176-L181 so by the same reasoning I suggest leaving the code as-is. Fair enough?

Reply:
These are secondary methods. `read` uses UTF-8 so everything else should be updated accordingly.

Reply:
commit 14d980c437 FTR

Reply:
> `read` uses UTF-8 so everything else should be updated accordingly.

Pre-existing issue, but never mind that. Done.

Reply:
Actually I am not comfortable taking on this risk changing the behavior of an existing public API. Reverted.

Reply:
Then at least leave comments noting that the behavior of methods in this class is inconsistent. Not even sure if these non-`read` methods are used.

Reply:
Opted for a more conservative approach of deprecating the methods that use the default charset in favor of non-deprecated versions that take an explicit charset.

Reply:
> read uses UTF-8 so everything else should be updated accordingly.

as does `write` and `lines` so these `head`/`tail` are bogus as is the comment on `fastTail` as the only possibility is for this to every be a `UTF-8` encoded text file?

Thus forcing users to specify a charset for `tail`/`head` to be "correct" when the only correct possibility is "UTF-8" for the encoding of the text in the file seems wrong to me, and the mere existance of methods that take a charset would be a flag (and I note that some already exist!).

Besides that - this code is wrong as it is ignoring the passed in `Charset` and using `Charset.defaultCharset()`

there only seems to be a couple of usages of [`fastTail`](https://github.com/search?q=org%3Ajenkinsci+fastTail&type=code) - 
 `junit` plugin seems to be using this for files that are not written by textFile but existing files.  That seems like an abuse to to me, and changing behaviour here without first fixing that would likely be a regression.

`usage-in-plugins` reports just a few callers of these methods - so it would seem to be better to fix those (if they are similar to junit), and then not take a charset at all and make TextFile always be `UTF-8`?

>flaky-test-handler:1.2.0
Methods
hudson/util/TextFile#fastTail(I)Ljava/lang/String;
hudson/util/TextFile#head(I)Ljava/lang/String;

>junit:1.53
Methods
hudson/util/TextFile#fastTail(I)Ljava/lang/String;
hudson/util/TextFile#head(I)Ljava/lang/String;

>mentor-questa-vrm:1.13
Methods
hudson/util/TextFile#fastTail(I)Ljava/lang/String;
hudson/util/TextFile#head(I)Ljava/lang/String;